flashcards.go

```diff
@@ -0,0 +1,202 @@
+package main
+
+import (
+	//"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	tba "gopkg.in/tucnak/telebot.v2" //telegram bot api
+)
+
+func writeFlashcards(fc flashcards, ioLogger *log.Entry) error {
+	fcJSON, err := json.Marshal(fc)
+	if err != nil {
+		ioLogger.Error("Could not encode flashcards")
+		return err
+	}
+
+	err = ioutil.WriteFile(flashcardsFileName, fcJSON, 0644)
+	if err != nil {
+		ioLogger.Error("Could not write file")
+		return err
+	}
+	return err
+}
+
+func addFlashcard(fc flashcards, chatID chatid, out chan msg, in chan string, state chan chatid) {
+
+	chatLogger := generateDialogLogger(chatID)
+
+	ioLogger := generateIoLogger(flashcardsFileName, "addFlashcard")
+
+	t, err := dialog(out, chatID, "Podaj temat", in)
+	if err != nil {
+		chatLogger.Info("Dialog ended unsuccessfully")
+		state <- chatID
+		return
+	}
+	top := topic(t)
+
+	term, err := dialog(out, chatID, "Podaj pojecie", in)
+	if err != nil {
+		chatLogger.Info("Dialog ended unsuccessfully")
+		state <- chatID
+		return
+	}
+
+	if _, ok := fc[chatID][top][term]; ok {
+		out <- msg{chatID, "Fiszka juz istnieje, edytuj za pomoca /edytujfiszke"}
+		state <- chatID
+		return
+	}
+
+	definition, err := dialog(out, chatID, "Podaj definicje", in)
+	if err != nil {
+		chatLogger.Info("Dialog ended unsuccessfully")
+		state <- chatID
+		return
+	}
+
+	if fc[chatID] == nil {
+		fc[chatID] = make(map[topic]flashcard)
+	}
+
+	if fc[chatID][top] == nil {
+		fc[chatID][top] = make(flashcard)
+	}
+
+	fc[chatID][top][term] = definition
+
+	err = writeFlashcards(fc, ioLogger)
+	if err != nil {
+		out <- msg{chatID, "Wystapil problem, moga wystapic problemy z tym terminem w przyszlosci, skontaktuj sie z administratorem"}
+		state <- chatID
+		return
+	}
+
+	out <- msg{chatID, "Dodano fiszke"}
+	state <- chatID
+
+}
+
+func displayFlashcard(fc flashcards, m *tba.Message, output chan msg) {
+
+	chatID := chatid(m.Chat.ID)
+
+	if m.Text == "/fiszka" {
+		output <- msg{chatID, "Podaj pojecie po spacji"}
+		return
+	}
+
+	term := strings.ReplaceAll(m.Text, "/fiszka ", "")
+	flashcardFound := false
+
+	for top, val := range fc[chatID] {
+		if definition, ok := val[term]; ok {
+			flashcardFound = true
```

Comment:
Aktualnie wypisujesz kilka znalezionych definicji, wydajniej byłoby je znaleźć wszystkie i później wysłać jedną wiadomość z przygotowaną treścią (przy bardziej skomplikowanym formacie, zamiast sklejania stringów możesz rozważyć `text/template` albo `fmt.Sprintf`).

A jeśli chcesz wypisywać jedną definicję, to pominąłbym tę zmienną `flashcardFound` i dodał tutaj `return`.

Reply:
Ewentualnie, co jest bardziej ciekawym ćwiczeniem niż potrzebną tu optymalizacją, możesz zaimplementować łączenie wiadomości w gorutynie odbierającej wiadomości z tego kanału. Tj. przed wysłaniem może czekać np. 100 ms na kolejne wiadomości na ten sam kanał i wysłać wszystkie w jednej.